### PR TITLE
Ensure the experimental.vm.proxy settings are only followed on win32.

### LINKED ISSUE #5478: The settings validator isn't guarding the windows-specific experimental.virtualMachine fields
### Problem Description

Running this on a mac:

```
$ rdctl api -X PUT settings --body '{"experimental": {"virtualMachine": {"proxy": {"noproxy": ["c-stub", " ", "d-stub", "    ", "e-stub", ""] }}}, "version": 9}'
settings updated; no restart required
```

should give an error message that the `noproxy` field is for Windows only.

### PR DESCRIPTION
**Pull request by ericpromislow**
Fixes #5478

### CODE REVIEW

File: bats/tests/preferences/verify-settings.bats [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    wait_for_container_engine
+}
+
+@test 'complain about windows-specific vm settings' {
+    skip_on_windows
+    run rdctl api /settings
+    assert_success
+    proxyEnabled=$(jq .experimental.virtualMachine.proxy.enabled <<<"$output")
+    settingsTemplateStart='{ "version": '$(get_setting .version)', "experimental": { "virtualMachine": { "proxy":'
+    settingsTemplateEnd='} } }'
```
Comment (jandubois) [inline bats/tests/preferences/verify-settings.bats:18]: I think this is kind of hard to read. I would rather add a function

```shell
proxy_set() {
    local field=$1
    local value=$2

    run rdctl api settings -X PUT --body '{ "version": '$(get_setting .version)', "experimental": { "virtualMachine": { "proxy": { '$(json_string "$field")': $value}}}}'
    assert_failure
    assert_output --partial "$(printf $'Changing field "experimental.virtualMachine.proxy.%s" via the API isn\'t supported' $field)"
}
```

(Just hacked up in the web interface, so probably won't work as-is)

and then call it

```
@test ... {
    proxy_set enabled true
    proxy_set address "$(json_string "foo")"
    ...
}

File: bats/tests/preferences/verify-settings.bats [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    assert_failure
+    assert_output --partial $"Changing field \"experimental.virtualMachine.proxy.noproxy\" via the API isn't supported"
+}
+
+@test 'ignores echoing current vm settings' {
+    skip_on_windows
+    run rdctl api /settings
+    assert_success
+    proxyPart=$(jq .experimental.virtualMachine.proxy <<<"$output")
+    payload='{ "version": '$(get_setting .version)', "experimental": { "virtualMachine": { "proxy": '"$proxyPart"' } } }'
```
Comment (jandubois) [inline bats/tests/preferences/verify-settings.bats:50]:
```suggestion
    run jq_output .experimental.virtualMachine.proxy
    assert_success
    payload='{ "version": '$(get_setting .version)', "experimental": { "virtualMachine": { "proxy": '"$output"' } } }'
```

**Review by jandubois — CHANGES_REQUESTED** [review]
First round of feedback

File: bats/tests/preferences/verify-settings.bats [hunk 3]
```diff
@@ -0,0 +1,53 @@
+load '../helpers/load'
+
+@test 'initial factory reset' {
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply bats/tests/preferences/verify-settings.bats:3]: Except PR 5531 does some similar tests, with a test that runs only on Windows, Linux, and a subset of macOS versions/architectures. So it makes sense to have one file that verifies attempted settings, with specific requirements in each test.

File: bats/tests/preferences/verify-settings.bats (hunk 2 above)
Comment (ericpromislow) [reply bats/tests/preferences/verify-settings.bats]: Yes, and using `printfs` makes the bash more readable as well.

File: bats/tests/preferences/verify-settings.bats [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+@test 'start the app' {
+    start_container_engine
+    wait_for_container_engine
+}
+
+proxy_set() {
+    local field=$1
+    local value=$2
+
+    payload="$(printf '{ "version": %d, "experimental": { "virtualMachine": { "proxy": { "%s": %s }}}}' "$(get_setting .version)" "$field" "$value")"
```
Comment (jandubois) [inline bats/tests/preferences/verify-settings.bats:16]: The outer quotes are redundant:

```suggestion
    payload=$(printf '{ "version": %d, "experimental": { "virtualMachine": { "proxy": { "%s": %s }}}}' "$(get_setting .version)" "$field" "$value")
```

File: bats/tests/preferences/verify-settings.bats [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    proxy_set noproxy '["buffalo"]'
+}
+
+@test 'ignores echoing current vm settings' {
+    skip_on_windows
+    run rdctl api /settings
+    assert_success
+    run jq_output .experimental.virtualMachine.proxy
+    assert_success
+    payload="$(printf '{ "version": %s, "experimental": { "virtualMachine": { "proxy": %s } } }' "$(get_setting .version)" "$output")"
```
Comment (jandubois) [inline bats/tests/preferences/verify-settings.bats:47]: Again, no need for outer quotes

```suggestion
    payload=$(printf '{ "version": %s, "experimental": { "virtualMachine": { "proxy": %s } } }' "$(get_setting .version)" "$output")
```

File: bats/tests/preferences/verify-settings.bats (hunk 3 above)
[the remark this answers is not in the document]
Comment (jandubois) [reply bats/tests/preferences/verify-settings.bats:3]: I don't get your argument. There are literally **no** tests in this file if you run on Windows, so there is no need to spin up the application.

If you ever add tests in this file, that do run on Windows, then you can enable this again, but for now this is just wasting a couple of minutes time in a full BATS test on Windows.

**Review by jandubois — CHANGES_REQUESTED** [review]
Needs some changes...

File: bats/tests/preferences/verify-settings.bats [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    run rdctl api settings -X PUT --body "$payload"
+    assert_failure
+    assert_output --partial "Changing field \"experimental.virtualMachine.proxy.${field}\" via the API isn't supported"
+}
+
+@test 'complain about windows-specific vm settings' {
+    skip_on_windows
+    run rdctl api /settings
+    assert_success
+    run jq .experimental.virtualMachine.proxy.enabled <<<"$output"
```
Comment (jandubois) [inline bats/tests/preferences/verify-settings.bats:26]: Hmm, I added a comment here before, but Github seems to have swallowed it...

Use `jq_output` here; it is tidier and will throw an error if the field does not exist whereas plain `jq` will just return `null`:

```console
$ echo {} | jq .experimental.virtualMachine.proxy.enabled; echo $?
null
0
```

```suggestion
    run jq_output .experimental.virtualMachine.proxy.enabled
```

File: bats/tests/preferences/verify-settings.bats (hunk 3 above)
[the remark this answers is not in the document]
Comment (ericpromislow) [reply bats/tests/preferences/verify-settings.bats:3]: I've been working on multiple PRs in parallel, and both of them had created `bats/tests/preferences/verify-settings.bats` but sure, ok, I'll just move the `skip` tests back if and when I need to.

**Review by jandubois — APPROVED** [review]
Thanks, LGTM

**Pull request merged**